backend/GameManager.py

```diff
@@ -74,49 +71,82 @@ def is_empty(self, row, col):
     def is_adjacent_and_empty(self, current_row, current_col, target_row, target_col):
         """ Check if the target cell is adjacent to the current cell and is empty """
         if self.is_adjacent(current_row, current_col, target_row, target_col):
-            if (target_row, target_col) in self.current_moves:
+            if (target_row, target_col) in self.open_moves:
                 return True
         return False
-    
-    
+
+    def filter_empty_adjacent(self, adjacent_positions):
+        filtered_adjacent_positions = []
+        for row, col in adjacent_positions:
+            if self.board.check_position(row, col) == Color.EMPTY or self.board.check_position(row, col) == Color.VOID:
+                filtered_adjacent_positions.append((row, col))
+
+        return filtered_adjacent_positions
+
+    '''
+    the select_piece func should return the available empty adjacent positions to where the piece can be moved 
+    or throw exception if it should be selectable
+    '''
     def select_piece(self, row, col):
         if not self.placement_complete():
-            print("SelectionError -- Cannot select pieces during placement phase")
-            return False
-        
-        if self.turn == Turn.WHITE and self.board.check_position(row, col) == CellType.WHITE:
-            self.selected_piece = (row, col)
-            return True
-        
-        if self.turn == Turn.BLACK and self.board.check_position(row, col) == CellType.BLACK:
-            self.selected_piece = (row, col)
-            return True
-        
-        print("SelectionError -- Invalid piece selection")
-        return False
-        
+            raise Exception("SelectionError -- Cannot select pieces during placement phase")
+
+        cell = self.board.get_cell(row, col)
+        self.selected_piece = (row, col)
+
+        if self.turn == cell.get_state():
+            total_pieces_left = 9
+            if self.turn == Color.WHITE:
+                total_pieces_left = len(self.pieces.white_pieces)
+            elif self.turn == Color.BLACK:
+                total_pieces_left = len(self.pieces.black_pieces)
+
+            # calculate the available adjacent positions
+            if total_pieces_left > 3:
+                # if more than 3 pieces left, user can only move to empty adjacent positions
+                return self.filter_empty_adjacent(self.board.adjacent_positions_map[(row, col)])
```

Comment:
Instead of passing `row, col` could we pass `self.selected_piece`?

Reply:
I'm wondering if 'filter_empty_adjacent' should just take the coordinates and not a map of adjacent position. In 'filter_empty_adjacent' it finds the adjacent positions. So that function's responsibility should be 'find_empty_adjacent_positions' and build the coordinate array without including occupied cells to begin with.

Reply:
agreed. i made the changes in the new push